scrapy/utils/conf.py

```diff
@@ -1,36 +1,49 @@
 import os
 import sys
+import warnings
 from operator import itemgetter
 
 import six
 from six.moves.configparser import SafeConfigParser
 
+from scrapy.settings import BaseSettings
 from scrapy.utils.deprecate import update_classpath
 
 
-def build_component_list(base, custom, convert=update_classpath):
-    """Compose a component list based on a custom and base dict of components
-    (typically middlewares or extensions), unless custom is already a list, in
-    which case it's returned.
-    """
+def build_component_list(compdict, convert=update_classpath):
+    """Compose a component list from a { class: order } dictionary."""
 
     def _check_components(complist):
         if len({convert(c) for c in complist}) != len(complist):
             raise ValueError('Some paths in {!r} convert to the same object, '
                              'please update your settings'.format(complist))
 
-    if isinstance(custom, (list, tuple)):
-        _check_components(custom)
-        return type(custom)(convert(c) for c in custom)
-
     def _map_keys(compdict):
+        if isinstance(compdict, BaseSettings):
+            compbs = BaseSettings()
+            for k, v in six.iteritems(compdict):
+                prio = compdict.getpriority(k)
+                if compbs.getpriority(convert(k)) == prio:
+                    raise ValueError('Some paths in {!r} convert to the same '
+                                     'object, please update your settings'
+                                     ''.format(list(compdict.keys())))
+                else:
+                    compbs.set(convert(k), v, priority=prio)
+            return compbs
+        else:
+            _check_components(compdict)
+            return {convert(k): v for k, v in six.iteritems(compdict)}
+
+    if isinstance(compdict, (list, tuple)):
         _check_components(compdict)
-        return {convert(k): v for k, v in six.iteritems(compdict)}
+        return type(compdict)(convert(c) for c in compdict)
+    compdict = remove_none_values(_map_keys(compdict))
+    return [k for k, v in sorted(six.iteritems(compdict), key=itemgetter(1))]
+
 
-    compdict = _map_keys(base)
-    compdict.update(_map_keys(custom))
-    items = (x for x in six.iteritems(compdict) if x[1] is not None)
-    return [x[0] for x in sorted(items, key=itemgetter(1))]
+def remove_none_values(compdict):
```

Comment:
This utility function looks generic enough for scrapy.utils.conf - what about moving it to scrapy.utils.python?

Also, I expect `remove_none_values` it to remove values inplace. `without_none_values`? We can also extend it to handle iterables.